File: exercises/cpp/parser_code_review/parser.cpp
#include "parser.h"
#include <cctype>

Parser::Parser(std::string_view input) : input_(input) {}

ExprPtr Parser::parse() {
    skip_whitespace();
    auto expr = parse_expr();
    skip_whitespace();
    if (!at_end())
        throw ParseError("unexpected trailing characters at position " + std::to_string(pos_));
    return expr;
}

ExprPtr Parser::parse_expr() {
    skip_whitespace();
    if (at_end())
        throw ParseError("unexpected end of input");

    char c = peek();

    if (c == '-' || std::isdigit(static_cast<unsigned char>(c))) {
        // Could be a negative number or the start of a function name — digits confirm number
        if (c == '-') {
            // Look ahead: if next char is a digit, it's a negative number
            if (pos_ + 1 < input_.size() && std::isdigit(static_cast<unsigned char>(input_[pos_ + 1])))
                return parse_number();
            throw ParseError("unexpected '-' at position " + std::to_string(pos_));
        }
        return parse_number();
    }

    if (std::isalpha(static_cast<unsigned char>(c)))
        return parse_function_call();

    throw ParseError(std::string("unexpected character '") + c + "' at position " + std::to_string(pos_));
}

ExprPtr Parser::parse_number() {
    skip_whitespace();
    size_t start = pos_;
    if (peek() == '-')
        advance();
    if (at_end() || !std::isdigit(static_cast<unsigned char>(peek())))
        throw ParseError("expected digit at position " + std::to_string(pos_));
    while (!at_end() && std::isdigit(static_cast<unsigned char>(peek())))
        advance();
    std::string num_str(input_.substr(start, pos_ - start));
    return std::make_unique<NumberExpr>(std::stoi(num_str));
}

ExprPtr Parser::parse_function_call() {
    auto name = read_identifier();
    OpType op = identifier_to_op(name);

    skip_whitespace();
    expect('(');

    auto left = parse_expr();

    skip_whitespace();
    expect(',');

    auto right = parse_expr();

    skip_whitespace();
    expect(')');

    return std::make_unique<BinaryOpExpr>(op, std::move(left), std::move(right));
}

void Parser::skip_whitespace() {
    while (!at_end() && std::isspace(static_cast<unsigned char>(peek())))
        ++pos_;
}

char Parser::peek() const {
    if (at_end())
        throw ParseError("unexpected end of input");
    return input_[pos_];
}

char Parser::advance() {
    char c = peek();
    ++pos_;
    return c;
}

void Parser::expect(char c) {
    skip_whitespace();
    if (at_end())
        throw ParseError(std::string("expected '") + c + "' but reached end of input");
    if (peek() != c)
        throw ParseError(std::string("expected '") + c + "' but got '" + peek() + "' at position " + std::to_string(pos_));
    advance();
}

bool Parser::at_end() const {
    return pos_ >= input_.size();
}

std::string_view Parser::read_identifier() {
    skip_whitespace();
    size_t start = pos_;
    while (!at_end() && std::isalpha(static_cast<unsigned char>(peek())))
        advance();
    if (start == pos_)
        throw ParseError("expected identifier at position " + std::to_string(pos_));
    return input_.substr(start, pos_ - start);
}

OpType Parser::identifier_to_op(std::string_view name) const {
    if (name == "Sum")      return OpType::Sum;
    if (name == "Mult")     return OpType::Mult;
    if (name == "Div")      return OpType::Div;
    if (name == "Exponent") return OpType::Exponent;
    throw ParseError("unknown function: " + std::string(name));
}

File: exercises/csharp/parser-code-review/ExpressionEvaluator/Parser.cs
namespace ExpressionEvaluator;

public class ParseError : Exception
{
    public ParseError(string message) : base(message) { }
}

public class Parser
{
    private readonly ReadOnlyMemory<char> _input;
    private int _pos;

    public Parser(string input)
    {
        _input = input.AsMemory();
        _pos = 0;
    }

    public Expr Parse()
    {
        SkipWhitespace();
        var expr = ParseExpr();
        SkipWhitespace();
        if (!AtEnd)
            throw new ParseError($"unexpected trailing characters at position {_pos}");
        return expr;
    }

    private Expr ParseExpr()
    {
        SkipWhitespace();
        if (AtEnd)
            throw new ParseError("unexpected end of input");

        char c = Peek;

        if (c == '-')
        {
            if (_pos + 1 < _input.Length && char.IsDigit(_input.Span[_pos + 1]))
                return ParseNumber();
            throw new ParseError($"unexpected '-' at position {_pos}");
        }

        if (char.IsDigit(c))
            return ParseNumber();

        if (char.IsLetter(c))
            return ParseFunctionCall();

        throw new ParseError($"unexpected character '{c}' at position {_pos}");
    }

    private Expr ParseNumber()
    {
        SkipWhitespace();
        int start = _pos;
        if (Peek == '-')
            Advance();
        if (AtEnd || !char.IsDigit(Peek))
            throw new ParseError($"expected digit at position {_pos}");
        while (!AtEnd && char.IsDigit(Peek))
            Advance();
        var numStr = _input.Span[start.._pos];
        return new NumberExpr(int.Parse(numStr));
    }

    private Expr ParseFunctionCall()
    {
        var name = ReadIdentifier();
        var op = IdentifierToOp(name);

        SkipWhitespace();
        Expect('(');

        var left = ParseExpr();

        SkipWhitespace();
        Expect(',');

        var right = ParseExpr();

        SkipWhitespace();
        Expect(')');

        return new BinaryOpExpr(op, left, right);
    }

    private void SkipWhitespace()
    {
        while (!AtEnd && char.IsWhiteSpace(_input.Span[_pos]))
            _pos++;
    }

    private char Peek
    {
        get
        {
            if (AtEnd) throw new ParseError("unexpected end of input");
            return _input.Span[_pos];
        }
    }

    private char Advance()
    {
        char c = Peek;
        _pos++;
        return c;
    }

    private void Expect(char c)
    {
        SkipWhitespace();
        if (AtEnd)
            throw new ParseError($"expected '{c}' but reached end of input");
        if (Peek != c)
            throw new ParseError($"expected '{c}' but got '{Peek}' at position {_pos}");
        Advance();
    }

    private bool AtEnd => _pos >= _input.Length;

    private ReadOnlySpan<char> ReadIdentifier()
    {
        SkipWhitespace();
        int start = _pos;
        while (!AtEnd && char.IsLetter(_input.Span[_pos]))
            _pos++;
        if (start == _pos)
            throw new ParseError($"expected identifier at position {_pos}");
        return _input.Span[start.._pos];
    }

    private static OpType IdentifierToOp(ReadOnlySpan<char> name)
    {
        if (name.SequenceEqual("Sum")) return OpType.Sum;
        if (name.SequenceEqual("Mult")) return OpType.Mult;
        if (name.SequenceEqual("Div")) return OpType.Div;
        if (name.SequenceEqual("Exponent")) return OpType.Exponent;
        throw new ParseError($"unknown function: {name}");
    }
}
